### PR TITLE
[awattar] Initial contribution

### PR DESCRIPTION
**Pull request by Wolfgang1966**
Around one year ago I started developing a binding to retrieve energy prices for the German/Austrian electricity provider aWATTar. Find the discussion at
https://community.openhab.org/t/awattar-binding-beta-and-discussion/110497

The goal of the binding is to enable aWATTar customers to make use of hours with lower energy prices for equipment that consumes a lot of energy but does not need to run all the time, e.g. wallboxes. 

A few people are using the binding already in their openHAB installation. They found a few bugs which were already corrected since then.

Before opening this PR I rebased my branch to the current main branch to avoid merge conflicts.

The latest version of the binding is available for download at 
https://home.klimt.de/owncloud/index.php/s/zKAsTWkxjG6RLyS

I did my best to follow the requirements for new bindings. However, as this is my first one, I may have overlooked things, If I did, I appreciate any comments.

### CODE REVIEW

**Comment by lolodomo** [comment]
@Wolfgang1966  : I mark this PR with label "work in progress" in relation to your PR tile. Let us know when your work is finished and ready for a review.

**Comment by Wolfgang1966** [comment]
Hi @lolodomo , thanks for bringing this up. I removed the "WIP" from the title, as I now managed to get rid of all warnings (at least when building the code locally). So I would appreciate a first review to get some feedback. As I wrote, this is the first binding I wrote from scratch and there may be fundamental things I failed to implement the way it should be. I assume the first review brings up those things and will require deeper rework which will take some time. Afterwards, there may be smaller detail work necessary, but I think this only makes sense as soon as the overall structure is finalized and stable.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java:247]: Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be debug. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBridgeHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBridgeHandler.java:104]: The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.

Please check all.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBridgeHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBridgeHandler.java:126]: See above. Use `scheduleWithDelay`.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarHandlerFactory.java
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarHandlerFactory.java:73]: This is already logged by the framework and can therefore be removed.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarHandlerFactory.java
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarHandlerFactory.java:85]: See above. Already logged.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarPriceHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarPriceHandler.java:89]: See above. Please check below in this file.

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/binding/binding.xml
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/binding/binding.xml:7]: The binding name doesn't need to be repeated in the description.

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties:1]: @lolodomo Shall this be part of the binding?

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml:11]: See https://www.openhab.org/docs/developer/bindings/config-xml.html also for min/max values.

```suggestion
			<parameter name="country" type="text" pattern="DE|AT">
```

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml:27]: Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml:34]: Did you test if this also works if daylight saving time is active/not active?

**Review by fwolter — CHANGES_REQUESTED** [review]
Thanks for your contribution! I reviewed your code and here is my feedback.

You could use `mvn license:format` in the root directory of this repo to update the copyright year.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

Please add your binding to `bom\openhab-addons\pom.xml`.

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties
Comment (lolodomo) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties:1]: No, it is better to translate in Crowdin. 
But if the translation was fully done the way, as it is a new binding, an exception could be done : just after the merge, someone should take care to load the German transitions in Crowdin.

**Comment by Wolfgang1966** [comment]
@fwolter @lolodomo  Finally found some time to work on your comments. Hope to be able to push an update within the next days. Sorry for the delay.

File: bundles/org.openhab.binding.awattar/README.md
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/README.md]: Changed.

File: bundles/org.openhab.binding.awattar/README.md
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/README.md:90]: Changed

File: bundles/org.openhab.binding.awattar/src/main/feature/feature.xml
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/feature/feature.xml]: Removed

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/aWATTarUtil.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/aWATTarUtil.java]: changed to use TimeZoneProvider

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/binding/binding.xml
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/binding/binding.xml]: Changed

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties]: I removed the name from the properties file

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml]: Changed labels

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml]: It works i both Winter Time and Summer Time

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml]: Changed

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/aWATTarPrice.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/aWATTarPrice.java]: All changed

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/connection/aWATTarConfigurationException.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/connection/aWATTarConfigurationException.java]: Removed exception totally.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/connection/aWATTarConnectionException.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/connection/aWATTarConnectionException.java]: Removed Exception totally

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java]: removed

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java]: removed most of the trace logs. Only kept a few which from my point of view are helpful in case I need to support a user of the binding.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java]: Replaced most checks by XML config, except the last one which dynamically depends on another config value.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java]: At this point I need to keep it because the binding should act exactly at minute borders. This is not possible with scheduleWithFixedDelay without requiring further logic to adapt to the "drift" that would come with this approach. At least I was able to change one of the three refreshers I have.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java]: Done-

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBestpriceHandler.java]: See above: I removed most of the logging statements.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBridgeHandler.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBridgeHandler.java]: Checked and removed. I hope the few that remain are ok.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBridgeHandler.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarBridgeHandler.java]: Here I changed it.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarHandlerFactory.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarHandlerFactory.java]: Removed

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarPriceHandler.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/aWATTarPriceHandler.java]: This one is still necessary to meet the minute boundaries.

**Comment by Wolfgang1966** [comment]
Hi @fwolter and @lolodomo , I now did the requested changes and also the checkstyle report, with the following exception:

At two places I need the ScheduleAtFixedRate because the actions need to happen at minute boundaries. Otherwise I would have to schedule much more often than once per minute, which would most probably cause much more load than the more efficient scheduler might save.

Is there any way to avoid the warning in such cases?

**Comment by Wolfgang1966** [comment]
I just saw that the build fails because of the year in the header line. However, this is the line I got with
```
mvn license:format
``` 

and with "2022" the checkstyle report gives errors. Do I need to rebase the whole PR to a newer openHAB version to get rid of this problem?

**Review by fwolter — CHANGES_REQUESTED** [review]
To make the binding compile, you need to rebase your branch and set the version to the current snapshot: `3.3.0-SNAPSHOT`.

Here are the commands for rebasing your branch:

If not already done, add the upstream openHAB addon repo as a remote to your local repo and fetch it:
```
git remote add upstream https://github.com/openhab/openhab-addons.git
git fetch upstream
```
The remotes (`git remote --verbose`) should look like this, now:
```
origin  https://github.com/[your name]/openhab-addons.git (fetch)
origin  https://github.com/[your name]/openhab-addons.git (push)
upstream        https://github.com/openhab/openhab-addons.git (fetch)
upstream        https://github.com/openhab/openhab-addons.git (push)
```
Then, you can rebase your PR's branch onto `main`:
```
git rebase upstream/main
```
Finally force-push the rebased branch to this PR's branch:
```
git push origin [your branch name of this PR] --force-with-lease
```
**DON'T** use merge! This can clutter the Git history and make the PR unusable.

**Comment by Wolfgang1966** [comment]
Branch rebased.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/AwattarHandlerFactory.java
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/AwattarHandlerFactory.java:62]: Seems like you don't need `i18nProvider` and `localeProvider` anymore.

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/binding/binding.xml
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/binding/binding.xml:7]:
```suggestion
	<description>Hourly Electricity Prices for Germany and Austria.</description>
```

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml:12]: The binding name is redundant and can therefore be removed.
```suggestion
		<label>Hourly Price</label>
```

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties
Comment (fwolter) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties]: It was about the entire file. As you already did all translations in the properties file, I suggest you to add the name property again and after this has been merged, you ask a maintainer to load your whole translation into Crowdin.

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/AwattarBestpriceHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/AwattarBestpriceHandler.java:109]: This empty line needs to be removed to satisfy checkstyle.

Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/AwattarBestpriceHandler.java]: Removed it, however, There was no warning when I built the binding locally. I will double check the build result after push.

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml]: Changed

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/binding/binding.xml
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/binding/binding.xml]: Changed, thanks :-)

File: bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/AwattarHandlerFactory.java
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/java/org/openhab/binding/awattar/internal/handler/AwattarHandlerFactory.java]: You are right, removed them

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Wolfgang1966) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml]: Also found no checkstyle warning in the chenkins builds. Is there a special mvn goal for that? I use

```
mvn spotless:apply clean build
```

**Comment by Wolfgang1966** [comment]
Added the name property again, will take care of Crowdin after merge.

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/thing/thing-types.xml]: This particular point doesn't generate a warning. The warning about the empty line could be found in `target/code-analysis/report.html`.

**Review by fwolter — APPROVED** [review]
LGTM
Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

**Comment by Wolfgang1966** [comment]
Hi @fwolter , Github still claims the rebase of my PR as requested change. As I did rebase my branch, is here still anything open? At least the PR was not yet merged.

Yours

Wolfgang

File: bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties
Comment (fwolter) [reply bundles/org.openhab.binding.awattar/src/main/resources/OH-INF/i18n/awattar_de.properties:1]: @lolodomo @jlaur can you load the German translations into Crowdin?

**Comment by Wolfgang1966** [comment]
Contacted aWATTar for the Logo

**Comment by lolodomo** [comment]
Ok

**Comment by Wolfgang1966** [comment]
Is it possible to suppress the warning by some annotation or preceding comment in the code?

**Comment by fwolter** [comment]
Good idea! You could annotate the method with `@SuppressWarnings("PMD.AvoidScheduleAtFixedRateCheck")`.

**Pull request merged**
